ui/pods/poddialogs/stats.go

```diff
@@ -0,0 +1,482 @@
+				return
+			}
+		}
+		// stats table
+		if d.table.HasFocus() {
+			if event.Key() == tcell.KeyTab {
+				d.focusElement = podStatDialogFormFocus
+				setFocus(d)
+				return
+			}
+			if tableHanlder := d.table.InputHandler(); tableHanlder != nil {
+				tableHanlder(event, setFocus)
+				return
+			}
+		}
+
+	})
+}
+
+// Draw draws this primitive onto the screen.
+func (d *PodStatsDialog) Draw(screen tcell.Screen) {
+	if !d.display {
+		return
+	}
+	d.Box.DrawForSubclass(screen, d)
+	x, y, width, height := d.Box.GetInnerRect()
+	d.layout.SetRect(x, y, width, height)
+	d.layout.Draw(screen)
+
+}
+
+// SetRect set rects for this primitive.
+func (d *PodStatsDialog) SetRect(x, y, width, height int) {
+	dX := x + dialogs.DialogPadding
+	dY := y + dialogs.DialogPadding - 1
+	dWidth := width - (2 * dialogs.DialogPadding)
+	dHeight := height - (2 * (dialogs.DialogPadding - 1))
+	d.Box.SetRect(dX, dY, dWidth, dHeight)
+}
+
+// SetDoneFunc sets form cancel button selected function
+func (d *PodStatsDialog) SetDoneFunc(handler func()) *PodStatsDialog {
+	d.doneHandler = handler
+	cancelButton := d.form.GetButton(d.form.GetButtonCount() - 1)
+	cancelButton.SetSelectedFunc(handler)
+	return d
+}
+
+// SetPodsOptions sets pod drop down options
+func (d *PodStatsDialog) SetPodsOptions(options []PodStatsDropDownOptions) {
+	maxWidth := 0
+	d.podDropDownOptions = options
+	if len(options) == 0 {
+		return
+	}
+	podOptions := []string{"all"}
+	for i := 0; i < len(options); i++ {
+		item := options[i].ID
+		if options[i].Name != "" {
+			item = fmt.Sprintf("%s (%s)", item, options[i].Name)
+		}
+		if len(item) > maxWidth {
+			maxWidth = len(item)
+		}
+		podOptions = append(podOptions, item)
+	}
+	maxWidth = maxWidth + 11
+	d.controlLayout.ResizeItem(d.podDropDown, maxWidth, 0)
+	d.podDropDown.SetOptions(podOptions, d.setStatsQueryPodIDs)
+	d.podDropDown.SetCurrentOption(0)
+}
+
+func (d *PodStatsDialog) query() {
+	opts := d.getStatsQueryOptions()
+	podStats, err := ppods.Stats(opts)
+	if err != nil {
+		log.Error().Msgf("pod stats dialog: query error: %v", err)
+		return
+	}
+	d.updateData(podStats)
+}
+
+func (d *PodStatsDialog) startStatsQueryLoop() {
+	log.Debug().Msgf("pod stats dialog: starting pod stats query loop")
+	go func() {
+		tick := time.NewTicker(d.queryRefreshInterval)
+		// initial query
+		d.query()
+		for {
+			select {
+			case <-tick.C:
+				d.query()
+			case <-d.doneChan:
+				log.Debug().Msgf("pod stats dialog: stats query loop stopped")
+				return
+			}
+		}
+	}()
+}
+
+func (d *PodStatsDialog) getStatsQueryOptions() *ppods.StatsOptions {
+	d.mu.Lock()
+	opts := d.statQueryOpts
+	d.mu.Unlock()
+	return opts
+}
+
+func (d *PodStatsDialog) setStatsQueryPodIDs(name string, index int) {
+	if index == -1 {
+		return
+	}
+	d.mu.Lock()
+	if index == 0 {
+		d.statQueryOpts.IDs = d.getAllPodIDs()
+	} else {
+		d.statQueryOpts.IDs = []string{
+			d.podDropDownOptions[index-1].ID,
+		}
+	}
+	defer d.mu.Unlock()
```

Comment:
Why isn't the defer right after the Lock?  Of if all you want to lock is the section between the section above, why use defer?

Reply:
Thanks for review, correct , it should be after Lock.